docs/doc/60-contributing/03-rfcs/20220512-versioned-config.md

```diff
@@ -0,0 +1,146 @@
+
+# Motivation
+
+While early birds are starting to deploy databend by themselves, it's time for us to establish some contracts between users. We should allow users to upgrade their deployments without breaking backward compatibility. In this RFC, we will focus on `config`.
+
+`config` I mentioned here including:
+
+- config file that is read by `databend-query` and `databend-meta`.
+- config env that read by `databend-query` and `databend-meta`
+- application args that are accepted by `databend-query` and `databend-meta`
+- protobuf messaged that generated by `databend-query` (stored inside `databend-meta`)
+
+Out of scope:
+
+- Tools like `fuzz` and `metactl` are not covered by this RFC.
+- Command-line UX of `databend-query` and `databend-meta` is another topic. We will not cover it in this RFC.
+- Config input/output by SQL/HTTP Rest API are not covered (for example, the output of table `system.config`)
+
+> For convenience, I will use `databend` to refer to `databend-query` and `databend-meta`.
+
+With this RFC, our users will upgrade their deployments without breaking. Old configs should always work along with the new implementations.
+
+# Guide-level explanation
+
+With this RFC, users will have a new field called `version` at the top of the config file. The value of the version is an integer and defaults to `0`. If the version is not specified, we will load it as` 0`.
+
+All config files, env, and args will be supported:
+
+- config file: `version=42`
+- config env: `export CONFIG_VERSION=42`
+- args: `--config-version=42`
+
+Suppose compatible changes happened as a new config entry was added. databend will make sure that the entry has a default value.
+
+Suppose incompatible changes happened, like config been removed/renamed/changed. databend increases the config version. The older version will still load by the specified version and be converted to the latest config internally.
+
+# Reference-level explanation
+
+Inside databend, we will split config into `inner` and `outer`:
+
+**`inner`**
+
+Config instances used inside databend. All logic **SHOULD** be implemented towards `inner` config.
+
+**`outer`**
+
+Config instances used as the front office of databend. They will have different versions and transform into an `inner` config. Other modules **SHOULD NOT** depend on `outer` config.
+
+Take `query` for example:
+
+The inner config of the query will be like this:
+
+```rust
+#[derive(Clone, Default, Debug, PartialEq, Serialize, Deserialize)]
+#[serde(default)]
+pub struct Config {
+    pub query: QueryConfig,
+    pub log: LogConfig,
+    pub meta: MetaConfig,
+    pub storage: StorageConfig,
+    pub catalog: HiveCatalogConfig,
+}
+```
+
+The outer config of the query will be like this:
+
+```rust
+#[derive(Clone, Default, Debug, PartialEq, Serialize, Deserialize, Parser)]
+#[clap(about, version, author)]
+#[serde(default)]
+pub struct ConfigV1 {
+    #[clap(long, short = 'c', default_value_t)]
+    pub config_file: String,
```

Comment:
Do we need to create a new version if we just add new fields into the old config ?

Reply:
No, we only need to bump versions if we are doing incompatible changes like `config removed/renamed/changed`

Reply:
we can also consider the DEPRECATED approach when we need removing fields. just leave some time to users to remove the deprecated fields.

IMO, the config version change mostly happens with some developer-experience-driven redesign.

Reply:
Add a new `Config Maintenance` section to describe the actions we should take while updating config.

Reply:
Addressed via https://github.com/datafuselabs/databend/pull/5324#issuecomment-1125794431